### DISCUSSION TITLE
Ubuntu curl unable to get local issuer certificate

### DISCUSSION THREAD

**Discussion by engimatic** [curl/curl#10806] [Q&A] [answered]
When I use root user,and curl https，I success.When I use non root user,
`curl -vvv --ciphers DEFAULT@SECLEVEL=1 HTTPS://`  ，I got the ssl certificate problem:unable to get local issuer certificate.
Then I run `curl -vvv --ciphers DEFAULT@SECLEVEL=2 HTTPS://`, I got the  ssl certificate problem:EE certificate key too weak.

Why root and non root user encountered a different problem and how to reslove it?

**Comment by bagder** [comment, marked as answer]
"unable to get local issuer certificate" *usually* means that the server did not provide the intermediate certificate in the TLS handshake. Ie a server config error. You can verify that by pointing a TLS checker on the site, like https://www.ssllabs.com/ssltest/

**Reply by engimatic** [reply]
I test a lot of website.And something important is that, why root user success?

**Reply by bagder** [reply]
Ah didn't pay attention to that detail. That seems hard to understand. Perhaps because something in your system does not have the right permission for that user?

**Reply by engimatic** [reply]
All I know related permission is `/etc/ssl/certs` and `/usr/local/share/ca-certificates` ,but both have read access.

**Reply by engimatic** [reply]
I think it really should be a matter of permissions.But I donot know which permissions should I authorized.

**Comment by jay** [comment]
Maybe there is a difference in the curl configuration for your root user and your regular user. Run verbose curl -v and compare the certificate locations:

~~~
*  CAfile: /etc/ssl/certs/ca-certificates.crt
*  CApath: none
~~~

The CA certificate should be in one of those locations, either in CAfile (usually a certificate bundle, multiple CA certs in one file) or as a separate file in the directory CApath.

SECLEVEL=1 is a separate issue. Probably you are attempting to communicate with a server using an older TLS protocol (1.0,1.1) which is why you can't use SECLEVEL=2.

**Reply by engimatic** [reply]
nonroot user,curl -v:
`*   Trying 106.122.248.41:443...
* TCP_NODELAY set
* Connected to goproxy.cn (106.122.248.41) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/certs/ca-certificates.crt
  CApath: /etc/ssl/certs
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
* TLSv1.3 (IN), TLS handshake, Certificate (11):
* TLSv1.3 (OUT), TLS alert, bad certificate (554):
* SSL certificate problem: EE certificate key too weak
* Closing connection 0
curl: (60) SSL certificate problem: EE certificate key too weak
More details here: https://curl.haxx.se/docs/sslcerts.html
`



root user:
`* TCP_NODELAY set
* Connected to goproxy.cn (106.122.248.41) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/certs/ca-certificates.crt
  CApath: /etc/ssl/certs
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
* TLSv1.3 (IN), TLS handshake, Certificate (11):
* TLSv1.3 (IN), TLS handshake, CERT verify (15):
* TLSv1.3 (IN), TLS handshake, Finished (20):
* TLSv1.3 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.3 (OUT), TLS handshake, Finished (20):
* SSL connection using TLSv1.3 / TLS_AES_256_GCM_SHA384
* ALPN, server accepted to use h2
* Server certificate:
*  subject: CN=*.goproxy.cn
*  start date: Jul 29 00:00:00 2022 GMT
*  expire date: Aug 28 23:59:59 2023 GMT
*  subjectAltName: host "goproxy.cn" matched cert's "goproxy.cn"
*  issuer: C=CN; O=TrustAsia Technologies, Inc.; CN=TrustAsia RSA DV TLS CA G2
*  SSL certificate verify ok.
* Using HTTP2, server supports multi-use
* Connection state changed (HTTP/2 confirmed)
* Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
* Using Stream ID: 1 (easy handle 0x55d456f619b0)
`

**Reply by jay** [reply]
> * curl: (60) SSL certificate problem: EE certificate key too weak

Yes, you mentioned that you see that with `--ciphers DEFAULT@SECLEVEL=2` but what is the verbose output of `--ciphers DEFAULT@SECLEVEL=1` ?

**Reply by engimatic** [reply]
DEFAULT@SECLEVEL=1    error is  "unable to get local issuer certificate"

**Reply by jay** [reply]
Yes... but what is the verbose output when using that option? If you think there is a permissions issue you can try monitoring files that are opened and comparing `strace -e trace=file curl ...`

**Comment by frehberg** [comment]
As for the read permissions,  the certiificates in /etc/ssl/certs/ are using sym-links to other locations, please check read-permissions for all locations with your "user" the following way

```bash
find -L /etc/ssl/certs/ \! -perm /o+r
```

**Reply by engimatic** [reply]
return nothing
